drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,389 @@
+
+static void fill_extended_id(uint32_t id, uint8_t *bytebuf)
+{
+    bytebuf[0] = (uint8_t) ((id & 0x1FE00000UL) >> 21); /* T/RXBnSIDH */
+    bytebuf[1] = (uint8_t) ((id & 0x001C0000UL) >> 13)
+        | (uint8_t) ((id & 0x00030000UL) >> 16) | 0x08; /* T/RXBnSIDL */
+    bytebuf[2] = (uint8_t) ((id & 0x0000FF00UL) >> 8);  /* T/RXBnEID8 */
+    bytebuf[3] = (uint8_t) (id & 0x000000FFUL);         /* T/RXBnEID0 */
+}
+
+int mcp2515_send(candev_mcp2515_t *dev, const struct can_frame *frame, int mailbox)
+{
+    uint8_t prio = 1; /* TODO: adjust priority */
+    uint8_t outbuf[BUFFER_SIZE];
+    uint8_t ctrl;
+    
+    struct can_frame framebuf;
+
+    if (frame->can_dlc > CAN_MAX_DLEN) {
+        return -1;
+    }
+
+    framebuf.can_id = frame->can_id;
+    framebuf.can_dlc = frame->can_dlc;
+    for(int i = 0; i < framebuf.can_dlc; i++) {
+        framebuf.data[i] = frame->data[i];
+    }
+
+    //TODO: for speedup, remove this check
+    mcp2515_spi_read(dev, TX_CTRL(mailbox), &ctrl, 1);
+    if (ctrl & MCP2515_TXBCTRL_TXREQ) {
+        DEBUG("Mailbox in use, TXB%dCTRL: 0x%02x\n", mailbox, ctrl);
+        return -1;
+    }
+
+    if(framebuf.can_id > CAN_SFF_MASK){
+        framebuf.can_id |= CAN_EFF_FLAG;
+    }
+
+    if ((framebuf.can_id & CAN_EFF_FLAG) == CAN_EFF_FLAG) {
+        fill_extended_id(framebuf.can_id, outbuf);
+    }
+    else {
+        fill_standard_id(framebuf.can_id, outbuf);
+    }
+
+    outbuf[4] = framebuf.can_dlc;
+    memcpy(outbuf+5, framebuf.data, framebuf.can_dlc);
+
+    /* set mailbox priority */
+    mcp2515_spi_write(dev, TX_CTRL(mailbox), &prio, 1);
+
+    mcp2515_spi_write_txbuf(dev, mailbox, outbuf, 5 + framebuf.can_dlc);
+    mcp2515_enable_irq(dev, MCP2515_CANINTE_TX0IE << mailbox);
+    mcp2515_spi_rts(dev, mailbox);
+
+    return mailbox;
+}
+
+int mcp2515_receive(candev_mcp2515_t *dev, struct can_frame *frame, int mailbox)
+{
+    uint8_t inbuf[BUFFER_SIZE];
+
+    mcp2515_spi_read_rxbuf(dev, mailbox, inbuf, BUFFER_SIZE);
+
+    /* extended id */
+    if (inbuf[1] & MCP2515_RX_IDE) {
+        frame->can_id = ((uint32_t)inbuf[0] << 21) +
+           (((uint32_t)inbuf[1] & 0xE0) << 13) +
+           (((uint32_t)inbuf[1] & 0x03) << 16) +
+           ((uint32_t)inbuf[2] << 8) +
+           inbuf[3];
+        frame->can_id |= CAN_EFF_FLAG;
+    }
+
+    /* standard id */
+    else {
+        frame->can_id = ((uint32_t)inbuf[0] << 3) + (((uint32_t)inbuf[1] & 0xE0) >> 5);
+    }
+
+    frame->can_dlc = inbuf[4];
+    memcpy(frame->data, inbuf + 5, frame->can_dlc);
+    return mailbox;
+}
+
+int mcp2515_abort(candev_mcp2515_t *dev, int mailbox)
+{
+    return mcp2515_spi_bitmod(dev, TX_CTRL(mailbox), MCP2515_TXBCTRL_TXREQ, 0);
+}
+
+enum mcp2515_mode mcp2515_get_mode(candev_mcp2515_t *dev)
+{
+    enum mcp2515_mode result = MODE_UNKNOWN;
+    uint8_t mode;
+
+    int res = mcp2515_spi_read(dev, MCP2515_CANSTAT, &mode, 1);
+    if (res == 0) {
+        //TODO: error handling of extra information in mode result
```

Comment:
No idea tbh. I'll have to check the datasheet again to figure out what can happen there. 
But to be honest, I have never encountered any unexpected behaviour that can be related to that. 
Not sure what to do with it. Maybe @vincent-d can shed some light on this?

Reply:
Obviously it would be better to do the TODO, however, I could live with a reference (RM and section number) to the datasheet explaining that more fine grained values can be gathered (if this is actually case)

Reply:
If I understand it correctly, this CANSTAT register holds two nibbles: OPMOD and ICOD. the former is read in that function to get the current mode. The latter holds information about interrupts. One of those flags is an "interrupt error" flag. 

I don't really see how this relates to getting the mode, so I think it's unrelated to that function. I removed the TODO for now, but don't hesitate to object to this decision.